datasets/shopping.py

```diff
@@ -125,41 +125,68 @@ def evaluate_correct(
+        2. Another third depends on those items' quantities matching the expected values.
+        3. The last third is given if there are no hallucinated items.
+        """
+        score = 0
+        max_score = self.item_changes
         num_correct = 0
-        penalties = 0
-        errors = []
+        present = []
+        hallucinated = []
+        reasoning = []
+
+        expected_names = []
+        expected_items = {}
+        for a in expected_answers:
+            expected_names.append(a[0])
+            expected_items[a[0]] = a[1]
+
+        # Check response format
         try:
             answer_items = sanitize_and_parse_json(responses[0])
-            expected_items = {}
-            expected_names = []
-            for a in expected_answers:
-                expected_names.append(a[0])
-                expected_items[a[0]] = a[1]
+            assert isinstance(answer_items, list)
             for item in answer_items:
-                name = item["item"].lower()
-                matched, key = match_plural(name, expected_names)
-                if matched:
-                    if item["quantity"] == expected_items[key]:
-                        num_correct += 1
-                    else:
-                        errors.append(f"Wrong quantity for {name}: {item['quantity']} vs {expected_items[key]}.")
-                else:
-                    penalties += 1
-                    errors.append(f'Unknown item {name}.')
-        except (TypeError, ValueError, JSONDecodeError):
-            logging.exception(f"Response not in correct format")
-            num_correct = 0
-        if num_correct < max_score:
-            errors.append(f"{max_score - num_correct} items were not found in the response.")
-
-        num_correct = max(num_correct - penalties, 0)
-        if num_correct == max_score:
-            reasoning = "All items and quantities match."
-        else:
-            reasoning = "\n".join(errors)
-        return num_correct, max_score, [reasoning]
+                assert isinstance(item["item"], str)
+                assert isinstance(item["quantity"], int)
+        except (JSONDecodeError, ValueError, KeyError, AssertionError) as exc:
+            msg = f"Response not in correct format ({repr(exc)}):\n{responses[0]}"
+            logging.exception(msg)
+            return score, max_score, [msg]
+
+        # Evaluate
+        for item in answer_items:
+            name = item["item"].lower()
+            matched, key = match_plural(name, expected_names)
+            if not matched:
+                hallucinated.append(name)
+                continue
+            present.append(name)
+            if item["quantity"] == expected_items[key]:
+                num_correct += 1
+            else:
+                reasoning.append(f"Wrong quantity for {name}: {item['quantity']} vs {expected_items[key]}.")
+
+        score += len(present) / len(expected_answers)
```

Comment:
This could theoretically lead to an unbounded score if I hallucinate a lot of items. Maybe constrain it to 1?

Reply:
Actually not since the addition to present is after the hallucination check. Maybe rename the variable to `present_and_correct` or something?